common/math/geometry/src/solver/polynomial_solver.cpp

```diff
@@ -12,176 +12,212 @@
+   */
+  const auto tschirnhaus_transformation = [](const auto a, const auto b, const auto c) {
+    /// @note The first element of the return value is q, the second element is r
+    return std::tuple<double, double>((a * a - 3 * b) / 9, (a * (2 * a * a - 9 * b) + 27 * c) / 54);
+  };
+
+  const auto solve_without_limit =
+    // clang-format off
+    [this, a](const auto q, const auto r) -> std::vector<std::complex<double>> {
+    // clang-format on
+    if (const double q3 = q * q * q; r * r <= (q3 + tolerance)) {
+      /**
+       * @note If 3 real solutions are found.
+       * The URL specified in @sa is a reference material for developers who wish to follow the formulas,
+       * and the code that exists in the material is not included in this library.
+       * @sa https://onihusube.hatenablog.com/entry/2018/10/08/140426
+       */
+      const double t = std::acos(std::clamp(r / std::sqrt(q3), -1.0, 1.0));
+      return {
+        // clang-format off
+        std::complex<double>(-2 * std::sqrt(q) * std::cos( t                                             / 3) - a / 3, 0),
+        std::complex<double>(-2 * std::sqrt(q) * std::cos((t + boost::math::constants::two_pi<double>()) / 3) - a / 3, 0),
+        std::complex<double>(-2 * std::sqrt(q) * std::cos((t - boost::math::constants::two_pi<double>()) / 3) - a / 3, 0)
+        // clang-format on
+      };
+    } else {
+      /// @note If imaginary solutions exist.
+      const double A = [r, q3]() {
+        const auto calculate_real_solution = [r, q3]() {
+          return -std::cbrt(std::abs(r) + std::sqrt(r * r - q3));
+        };
+        return r < 0 ? -1 * calculate_real_solution() : calculate_real_solution();
+      }();
+      const double B = isEqual(A, 0) ? 0 : q / A;
+      /// @note If the imaginary part of the complex almost zero, this equation has a multiple solution.
+      const double imaginary_part = 0.5 * std::sqrt(3.0) * (A - B);
+      return isEqual(imaginary_part, 0)
+               ? std::vector<std::complex<double>>({
+                   // clang-format off
+                    std::complex<double>(       (A + B) - a / 3, 0),
+                    std::complex<double>(-0.5 * (A + B) - a / 3, 0)
+                   // clang-format on
+                 })
+               : std::vector<std::complex<double>>({
+                   // clang-format off
+                    std::complex<double>(       (A + B) - a / 3,               0),
+                    std::complex<double>(-0.5 * (A + B) - a / 3, -imaginary_part),
+                    std::complex<double>(-0.5 * (A + B) - a / 3,  imaginary_part)
+                   // clang-format on
+                 });
+    }
+    /// @note No solutions are found.
+    return {};
+  };
+  return std::apply(solve_without_limit, tschirnhaus_transformation(a, b, c));
 }
 
-double PolynomialSolver::root3(double x) const
+auto PolynomialSolver::isEqual(double value0, double value1) const -> bool
```

Comment:
The name `isEqual` is inappropriate, since this operation does not evaluate the two floating point numbers to be exactly equal. I suggest `isApproximatelyEqualTo` as a more appropriate name.